internal/locate/region_request.go

```diff
@@ -1656,14 +1657,17 @@ func (s *RegionRequestSender) onRegionError(bo *retry.Backoffer, ctx *RPCContext
 			zap.Uint64("safe-ts", regionErr.GetDataIsNotReady().GetSafeTs()),
 			zap.Stringer("ctx", ctx),
 		)
-		if !req.IsGlobalStaleRead() {
-			// only backoff local stale reads as global should retry immediately against the leader as a normal read
-			err = bo.Backoff(retry.BoMaxDataNotReady, errors.New("data is not ready"))
-			if err != nil {
-				return false, err
+		// stale reads should retry immediately against the leader as a normal read.
+		if s.replicaSelector != nil && s.replicaSelector.state != nil {
+			if af, ok := s.replicaSelector.state.(*accessFollower); ok && !af.resetStaleRead {
+				af.resetStaleRead = true
+				// always retry data-is-not-ready error.
+				return true, nil
 			}
 		}
-		return true, nil
+		// backoff other cases to avoid infinite retries.
+		err = bo.Backoff(retry.BoMaxDataNotReady, errors.New("data is not ready"))
+		return err == nil, err
```

Comment:
```suggestion
		err = bo.Backoff(retry.BoMaxDataNotReady, errors.New("data is not ready"))
		if err != nil {
		    return false, err
		}
		return true, nil
```

It's better to return `nil` value directly and follow the error processing convention of Golang, as the `nil`, `error`, and interface types in Golang have some design drawbacks.

Consider this:
```
func testError() *TestError {
   return nil
}
func test() error {
   return testError()
}

func main() {
    err := test()
    fmt.Println(err == nil) // The result is false.
}
```